third_party/xla_client/xla_util.cc

```diff
@@ -58,18 +75,13 @@ StatusOr<std::string> GetComputationHloText(const XlaComputation& computation) {
 }
 
 void ReportComputationError(
-    const Status& status,
-    absl::Span<const XlaComputation* const> computations,
+    const Status& status, absl::Span<const XlaComputation* const> computations,
     absl::Span<const Shape* const> output_shapes) {
   std::stringstream ss;
   for (size_t i = 0; i < computations.size(); ++i) {
     std::string hlo_text = GetComputationHloText(*computations[i]).ValueOrDie();
     MaybeSaveHloGraph(hlo_text, i);
-    ss << ">>> Dumping Computation " << i << "\n";
-    ss << hlo_text << "\n";
-    if (i < output_shapes.size() && output_shapes[i] != nullptr) {
-      ss << "OutputShape: " << *output_shapes[i] << "\n\n";
-    }
+    ss << MaybeDumpHloGraph(output_shapes, hlo_text, i);
```

Comment:
Maybe we can dump the first 10~20 lines and last 10~20 lines of the hlo by default? This way we know which graph caused the error.

Reply:
I just hope that we can minimize any noise for the user as it is very easy to give up on debugging any stack trace if there is a HLO dump as those pretty much mean nothing to the user. For example the first 5 and last 5 lines for a simple model like resnet50:
`
2021-02-10 03:53:38.327068: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98] >>> Dumping Computation 0
2021-02-10 03:53:38.327164: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98] HloModule SyncTensorsGraph.6698, input_output_alias={ {0}: (2, {}, may-alias), {1}: (29, {}, may-alias), {2}: (28, {}, may-alias), {3}: (27, {}, may-alias), {4}: (26, {}, may-alias), {5}: (45, {}, may-alias), {6}: (44, {}, may-alias), {7}: (43, {}, may-alias), {8}: (42, {}, may-alias), {9}: (41, {}, may-alias), {10}: (40, {}, may-alias), {11}: (39, {}, may-alias), {12}: (38, {}, may-alias), {13}: (37, {}, may-alias), {14}: (36, {}, may-alias), {15}: (35, {}, may-alias), {16}: (34, {}, may-alias), {17}: (33, {}, may-alias), {18}: (32, {}, may-alias), {19}: (31, {}, may-alias), {20}: (25, {}, may-alias), {21}: (24, {}, may-alias), {22}: (23, {}, may-alias), {23}: (22, {}, may-alias), {24}: (21, {}, may-alias), {25}: (60, {}, may-alias), {26}: (59, {}, may-alias), {27}: (58, {}, may-alias), {28}: (57, {}, may-alias), {29}: (56, {}, may-alias), {30}: (55, {}, may-alias), {31}: (54, {}, may-alias), {32}: (53, {}, may-alias), {33}: (52, {}, may-alias), {34}: (51, {}, may-alias), {35}: (50, {}, may-alias), {36}: (49, {}, may-alias), {37}: (48, {}, may-alias), {38}: (47, {}, may-alias), {39}: (46, {}, may-alias), {40}: (75, {}, may-alias), {41}: (74, {}, may-alias), {42}: (73, {}, may-alias), {43}: (72, {}, may-alias), {44}: (71, {}, may-alias), {45}: (70, {}, may-alias), {46}: (69, {}, may-alias), {47}: (68, {}, may-alias), {48}: (67, {}, may-alias), {49}: (66, {}, may-alias), {50}: (65, {}, may-alias), {51}: (64, {}, may-alias), {52}: (63, {}, may-alias), {53}: (62, {}, may-alias), {54}: (61, {}, may-alias), {55}: (90, {}, may-alias), {56}: (89, {}, may-alias), {57}: (88, {}, may-alias), {58}: (87, {}, may-alias), {59}: (86, {}, may-alias), {60}: (85, {}, may-alias), {61}: (84, {}, may-alias), {62}: (83, {}, may-alias), {63}: (82, {}, may-alias), {64}: (81, {}, may-alias), {65}: (80, {}, may-alias), {66}: (79, {}, may-alias), {67}: (78, {}, may-alias), {68}: (77, {}, may-alias), {69}: (76, {}, may-alias), {70}: (20, {}, may-alias), {71}: (19, {}, may-alias), {72}: (18, {}, may-alias), {73}: (17, {}, may-alias), {74}: (16, {}, may-alias), {75}: (105, {}, may-alias), {76}: (104, {}, may-alias), {77}: (103, {}, may-alias), {78}: (102, {}, may-alias), {79}: (101, {}, may-alias), {80}: (100, {}, may-alias), {81}: (99, {}, may-alias), {82}: (98, {}, may-alias), {83}: (97, {}, may-alias), {84}: (96, {}, may-alias), {85}: (95, {}, may-alias), {86}: (94, {}, may-alias), {87}: (93, {}, may-alias), {88}: (92, {}, may-alias), {89}: (91, {}, may-alias), {90}: (120, {}, may-alias), {91}: (119, {}, may-alias), {92}: (118, {}, may-alias), {93}: (117, {}, may-alias), {94}: (116, {}, may-alias), {95}: (115, {}, may-alias), {96}: (114, {}, may-alias), {97}: (113, {}, may-alias), {98}: (112, {}, may-alias), {99}: (111, {}, may-alias), {100}: (110, {}, may-alias), {101}: (109, {}, may-alias), {102}: (108, {}, may-alias), {103}: (107, {}, may-alias), {104}: (106, {}, may-alias), {105}: (135, {}, may-alias), {106}: (134, {}, may-alias), {107}: (133, {}, may-alias), {108}: (132, {}, may-alias), {109}: (131, {}, may-alias), {110}: (130, {}, may-alias), {111}: (129, {}, may-alias), {112}: (128, {}, may-alias), {113}: (127, {}, may-alias), {114}: (126, {}, may-alias), {115}: (125, {}, may-alias), {116}: (124, {}, may-alias), {117}: (123, {}, may-alias), {118}: (122, {}, may-alias), {119}: (121, {}, may-alias), {120}: (150, {}, may-alias), {121}: (149, {}, may-alias), {122}: (148, {}, may-alias), {123}: (147, {}, may-alias), {124}: (146, {}, may-alias), {125}: (145, {}, may-alias), {126}: (144, {}, may-alias), {127}: (143, {}, may-alias), {128}: (142, {}, may-alias), {129}: (141, {}, may-alias), {130}: (140, {}, may-alias), {131}: (139, {}, may-alias), {132}: (138, {}, may-alias), {133}: (137, {}, may-alias), {134}: (136, {}, may-alias), {135}: (15, {}, may-alias), {136}: (14, {}, may-alias), {137}: (13, {}, may-alias), {138}: (12, {}, may-alias), {139}: (11, {}, may-alias), {140}: (165, {}, may-alias), {141}: (164, {}, may-alias), {142}: (163, {}, may-alias), {143}: (162, {}, may-alias), {144}: (161, {}, may-alias), {145}: (160, {}, may-alias), {146}: (159, {}, may-alias), {147}: (158, {}, may-alias), {148}: (157, {}, may-alias), {149}: (156, {}, may-alias), {150}: (155, {}, may-alias), {151}: (154, {}, may-alias), {152}: (153, {}, may-alias), {153}: (152, {}, may-alias), {154}: (151, {}, may-alias), {155}: (180, {}, may-alias), {156}: (179, {}, may-alias), {157}: (178, {}, may-alias), {158}: (177, {}, may-alias), {159}: (176, {}, may-alias), {160}: (175, {}, may-alias), {161}: (174, {}, may-alias), {162}: (173, {}, may-alias), {163}: (172, {}, may-alias), {164}: (171, {}, may-alias), {165}: (170, {}, may-alias), {166}: (169, {}, may-alias), {167}: (168, {}, may-alias), {168}: (167, {}, may-alias), {169}: (166, {}, may-alias), {170}: (195, {}, may-alias), {171}: (194, {}, may-alias), {172}: (193, {}, may-alias), {173}: (192, {}, may-alias), {174}: (191, {}, may-alias), {175}: (190, {}, may-alias), {176}: (189, {}, may-alias), {177}: (188, {}, may-alias), {178}: (187, {}, may-alias), {179}: (186, {}, may-alias), {180}: (185, {}, may-alias), {181}: (184, {}, may-alias), {182}: (183, {}, may-alias), {183}: (182, {}, may-alias), {184}: (181, {}, may-alias), {185}: (210, {}, may-alias), {186}: (209, {}, may-alias), {187}: (208, {}, may-alias), {188}: (207, {}, may-alias), {189}: (206, {}, may-alias), {190}: (205, {}, may-alias), {191}: (204, {}, may-alias), {192}: (203, {}, may-alias), {193}: (202, {}, may-alias), {194}: (201, {}, may-alias), {195}: (200, {}, may-alias), {196}: (199, {}, may-alias), {197}: (198, {}, may-alias), {198}: (197, {}, may-alias), {199}: (196, {}, may-alias), {200}: (225, {}, may-alias), {201}: (224, {}, may-alias), {202}: (223, {}, may-alias), {203}: (222, {}, may-alias), {204}: (221, {}, may-alias), {205}: (220, {}, may-alias), {206}: (219, {}, may-alias), {207}: (218, {}, may-alias), {208}: (217, {}, may-alias), {209}: (216, {}, may-alias), {210}: (215, {}, may-alias), {211}: (214, {}, may-alias), {212}: (213, {}, may-alias), {213}: (212, {}, may-alias), {214}: (211, {}, may-alias), {215}: (240, {}, may-alias), {216}: (239, {}, may-alias), {217}: (238, {}, may-alias), {218}: (237, {}, may-alias), {219}: (236, {}, may-alias), {220}: (235, {}, may-alias), {221}: (234, {}, may-alias), {222}: (233, {}, may-alias), {223}: (232, {}, may-alias), {224}: (231, {}, may-alias), {225}: (230, {}, may-alias), {226}: (229, {}, may-alias), {227}: (228, {}, may-alias), {228}: (227, {}, may-alias), {229}: (226, {}, may-alias), {230}: (10, {}, may-alias), {231}: (9, {}, may-alias), {232}: (8, {}, may-alias), {233}: (7, {}, may-alias), {234}: (6, {}, may-alias), {235}: (255, {}, may-alias), {236}: (254, {}, may-alias), {237}: (253, {}, may-alias), {238}: (252, {}, may-alias), {239}: (251, {}, may-alias), {240}: (250, {}, may-alias), {241}: (249, {}, may-alias), {242}: (248, {}, may-alias), {243}: (247, {}, may-alias), {244}: (246, {}, may-alias), {245}: (245, {}, may-alias), {246}: (244, {}, may-alias), {247}: (243, {}, may-alias), {248}: (242, {}, may-alias), {249}: (241, {}, may-alias), {250}: (270, {}, may-alias), {251}: (269, {}, may-alias), {252}: (268, {}, may-alias), {253}: (267, {}, may-alias), {254}: (266, {}, may-alias), {255}: (265, {}, may-alias), {256}: (264, {}, may-alias), {257}: (263, {}, may-alias), {258}: (262, {}, may-alias), {259}: (261, {}, may-alias), {260}: (260, {}, may-alias), {261}: (259, {}, may-alias), {262}: (258, {}, may-alias), {263}: (257, {}, may-alias), {264}: (256, {}, may-alias), {265}: (5, {}, may-alias), {266}: (4, {}, may-alias) }
2021-02-10 03:53:38.327177: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98] 
2021-02-10 03:53:38.327183: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98] %max_F32.137 (lhs.138: f32[], rhs.139: f32[]) -> f32[] {
2021-02-10 03:53:38.327188: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98]   %lhs.138 = f32[] parameter(0)
2021-02-10 03:53:38.327194: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98]   %rhs.139 = f32[] parameter(1)
2021-02-10 03:53:38.327199: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98]   ROOT %maximum.140 = f32[] maximum(f32[] %lhs.138, f32[] %rhs.139)
2021-02-10 03:53:38.327203: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98] }
2021-02-10 03:53:38.327208: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98] 
2021-02-10 03:53:38.327212: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98] %ge_F32.152 (lhs.153: f32[], rhs.154: f32[]) -> pred[] {
2021-02-10 03:53:38.327217: E     463 tensorflow/compiler/xla/xla_client/xla_util.cc:98]   %lhs.153 = f32[] parameter(0)
`

I was thinking that any relevant info that fails execution would be present in the `tf::Status` returned by the session run (ex. unsupported types for an op). Do you think it'd be a big loss for our debugging ability if users didn't report that part? We wouldn't have the entire graph to debug anyways, so we'd need to ask them for a repro or dump of actual full failing graph?

Reply:
while I realize knowing which graph caused the error could be useful, I vote for clean output here. I personally don't recall ever looking at the graph output in stderr.